src/librustc/mir/mod.rs

```diff
@@ -948,10 +948,64 @@ pub struct VarDebugInfo<'tcx> {
     /// (see `LocalDecl`'s `source_info` field for more details).
     pub source_info: SourceInfo,
 
-    /// Where the data for this user variable is to be found.
-    /// NOTE(eddyb) There's an unenforced invariant that this `Place` is
-    /// based on a `Local`, not a `Static`, and contains no indexing.
-    pub place: Place<'tcx>,
+    /// How the contents of this user variable are represented.
+    pub contents: VarDebugInfoContents<'tcx>,
+}
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable, HashStable, TypeFoldable)]
+pub enum VarDebugInfoContents<'tcx> {
+    /// The user variable's data is entirely in one `Place`.
+    // NOTE(eddyb) there's an unenforced invariant that this `Place` is
```

Comment:
Can we add `debug_assert!()`s for these invariants?

Reply:
I mean, codegen already has a `span_bug!`, it's just that it's not enforced by types (and we don't have a MIR "validator" pass).